File: 2017/js/day-24.js
const fs = require('fs');
const input = fs.readFileSync('2017/input/day-24.input').toString();

const ports = input.split('\n').map(x => x.split('/').map(x => +x));

function getBridges(prev, ports) {
  const lastPort = prev[prev.length - 1];
  const candidates = ports.filter(port => lastPort[1] === port[0] || lastPort[1] === port[1]);
  if (candidates.length === 0) {
    return [prev];
  } else {
    let res = [];
    candidates.forEach(port => {
      const portIndex = ports.indexOf(port);
      const nextPorts = [...ports.slice(0, portIndex), ...ports.slice(portIndex + 1)];
      const adjustedPort = lastPort[1] === port[0] ? port : [port[1], port[0]];
      const nextPrev = [...prev, adjustedPort];
      res.push(...getBridges(nextPrev, nextPorts));
    });
    return res;
  }
}

const bridges = getBridges([[0, 0]], ports);
let max = 0;
let maxBridge = [];
bridges.forEach(bridge => {
  const sum = bridge.reduce((sum, x) => sum + x[0] + x[1], 0);
  if (bridge.length > maxBridge.length ||
    (bridge.length === maxBridge.length && sum > max)) {
    max = sum;
    maxBridge = bridge;
  }
});
console.log(max);

File: 2017/day-24.py
def get_bridges(prev, ports):
    lastPort = prev[-1]
    candidates = [port for port in ports if lastPort[1] == port[0] or lastPort[1] == port[1]]
    bridges = []
    if not candidates:
        bridges.append(prev)
    else:
        for port in candidates:
            portIndex = ports.index(port)
            nextPorts = [x for x in ports if x != port]
            nextPrev = [x for x in prev]
            nextPrev.append(port if lastPort[1] == port[    0] else [port[1], port[0]])
            nextBridges = get_bridges(nextPrev, nextPorts)
            for bridge in nextBridges:
                bridges.append(bridge)
    return bridges

input = open("day-24.input", "r").read().strip().split("\n")
ports = list(map(lambda x: [int(y) for y in x.split('/')], input))
bridges = get_bridges([[0,0]], ports)

maxVal = 0
maxBridge = []
for bridge in bridges:
    sumVal = sum(a+b for a, b in bridge)
    lengthVal = len(bridge)
    if (len(bridge) > len(maxBridge) or 
            (len(bridge) == len(maxBridge) and sumVal > maxVal)):
        maxVal = sumVal
        maxBridge = bridge
print(maxVal)
